order-service/src/main/java/com/innowise/orderservice/service/impl/OrderServiceImpl.java

```diff
@@ -0,0 +1,177 @@
+
+        return orderMapper.toResponseDtoList(retrievedOrders, getCustomersInfoOrFallbackMap(userIds));
+    }
+
+    @Override
+    @Transactional(readOnly = true)
+    public List<OrderResponseDto> getOrdersByStatuses(List<OrderStatus> statuses) {
+        List<Order> retrievedOrders = orderRepository.findOrdersByStatusIn(statuses);
+
+        List<Long> userIds = retrievedOrders.stream()
+                .map(Order::getUserId)
+                .distinct()
+                .toList();
+
+        return orderMapper.toResponseDtoList(retrievedOrders, getCustomersInfoOrFallbackMap(userIds));
+    }
+
+    @Override
+    @Transactional
+    public OrderResponseDto createOrder(Long userId, OrderRequestDto orderRequestDto) {
+        Order newOrder = orderMapper.toEntity(userId, orderRequestDto);
+
+        orderRequestDto.getItems().stream()
+                .map(this::convertToOrderItem)
+                .forEach(newOrder::addOrderItem);
+
+        Order createdOrder = orderRepository.save(newOrder);
+
+        return orderMapper.toDto(createdOrder, getCustomerInfoOrFallback(userId));
+    }
+
+    @Override
+    @Transactional
+    public OrderResponseDto updateOrder(Long orderId, OrderRequestDto orderRequestDto, Long userId) {
+        Order orderToUpdate = orderRepository.findOrderById(orderId)
+                .orElseThrow(() -> ResourceNotFoundException.orderNotFound(orderId));
+
+        if (!orderToUpdate.getUserId().equals(userId)) {
+            throw new AccessDeniedException("You don't have permission to access this order");
+        }
+
+        orderToUpdate.getOrderItems().clear();
+        orderRequestDto.getItems().stream()
+                .map(this::convertToOrderItem)
+                .forEach(orderToUpdate::addOrderItem);
+
+        Order updatedOrder = orderRepository.save(orderToUpdate);
+
+        return orderMapper.toDto(updatedOrder, getCustomerInfoOrFallback(userId));
+    }
+
+    @Override
+    @Transactional
+    public void cancelOrDeleteOrder(Long orderId, Long userId) {
+        Order orderToDelete = orderRepository.findOrderById(orderId)
+                .orElseThrow(() -> ResourceNotFoundException.orderNotFound(orderId));
+
+        if (!orderToDelete.getUserId().equals(userId) && !isUserAdmin()) {
+            throw new AccessDeniedException("You don't have permission to access this order");
+        }
+
+        if (isUserAdmin()) {
+            orderRepository.delete(orderToDelete);
+        } else {
+            if (orderToDelete.getStatus().equals(OrderStatus.CANCELLED)) {
+                throw OrderStatusException.orderIsAlreadyCancelled(orderId);
+            }
+            if (orderToDelete.getStatus().equals(OrderStatus.COMPLETED)) {
+                throw OrderStatusException.orderIsCompleted(orderId);
+            }
+            orderRepository.cancelOrderAsUser(orderId);
+        }
+    }
+
+    private boolean isUserAdmin() {
+        return SecurityContextHolder.getContext().getAuthentication()
```

Comment:
Consider using Spring Security's role hierarchy or role-based annotations for cleaner separation.